packages/test-utils/src/reporter.ts

```diff
@@ -873,48 +907,90 @@ class ReplayReporter<TRecordingMetadata extends UnstructuredMetadata = Unstructu
     }
   }
 
-  async enqueueUpload() {
-    if (this.recordingsToUpload.length) {
-      const recordings = [...this.recordingsToUpload];
-      this.recordingsToUpload = [];
-
-      this.pendingWork.push(
-        ...recordings
-          .filter(r => (this.filter ? this.filter(r) : true))
-          .map(r => this.uploadRecording(r))
-      );
+  private _storeUploadableTestResults(results: UploadableTestRunResult<TRecordingMetadata>[]) {
+    if (this._uploadStatusThreshold === "none") {
+      return;
     }
-  }
 
-  async onEnd(): Promise<PendingWork[]> {
-    debug("onEnd");
+    for (const result of results) {
+      if (result.result === "skipped") {
+        continue;
+      }
 
-    if (this.upload) {
-      let timeout = 2000;
-      if (process.env.REPLAY_UPLOAD_DELAY) {
-        const userTimeout = Number.parseInt(process.env.REPLAY_UPLOAD_DELAY);
-        if (!isNaN(userTimeout)) {
-          timeout = userTimeout;
-        }
-        debug("REPLAY_UPLOAD_DELAY value %d using %d", userTimeout, timeout);
+      let uploadableResults = this._uploadableResults.get(result.testId);
+      if (!uploadableResults) {
+        uploadableResults = [];
+        this._uploadableResults.set(result.testId, uploadableResults);
       }
+      uploadableResults.push(result);
 
-      await new Promise(resolve => setTimeout(resolve, timeout));
-      this.enqueueUpload();
+      if (result.result === "passed" || result.attempt >= result.maxAttempts) {
+        this._enqueueUploads(uploadableResults);
+      }
     }
+  }
 
-    if (this.pendingWork.length === 0) {
-      return [];
+  private _enqueueUploads(results: UploadableTestRunResult<TRecordingMetadata>[]) {
+    const latestResult = results[results.length - 1];
```

Comment:
nit: I know that this method is only ever called with at least one thing in the array, but it's probably still worth adding an assert to ensure it stays that way?